### PR TITLE
Download Tangle binary against specific hash

### PR DESCRIPTION
**Pull request by drewstone**

### CODE REVIEW

**Comment by drewstone** [comment]
@Serial-ATA I merged your branch in here btw.

**Comment by drewstone** [comment]
CC @shekohex if you're able to parse this error more quickly than us.

**Comment by drewstone** [comment]
Ah maybe forge isn't actually building it.

**Comment by drewstone** [comment]
Ok some new changes @tbraun96 @Serial-ATA @Tjemmmic @shekohex 
- Adding logging of the node to the testing using `RUST_LOG=debug`. If you have naming suggestions lmk, currently some output from the node (most of it) gets logged to stderr.
- There's a 3rd parameters to the `new_test_ext...` which specifies whether you want to use local or download tangle binary. This way we can test locally still if we need to add logging to pallets.

File: blueprint-test-utils/src/tangle/node.rs [hunk 1]
```diff
@@ -251,9 +323,41 @@ impl SubstrateNode {
 
         cmd.arg(format!("--rpc-port={}", self.ws_port));
         cmd.arg(format!("--port={}", self.p2p_port));
+
+        log::debug!(target: "gadget", "Restarting node with command: {:?}", cmd);
```
Comment (shekohex) [inline blueprint-test-utils/src/tangle/node.rs:327]:
```suggestion
        gadget_sdk::trace!("Restarting node with command: {cmd:?}");
```

File: blueprint-test-utils/src/tangle/node.rs [hunk 2]
```diff
@@ -264,23 +368,45 @@ impl SubstrateNode {
...
         let _ = self.kill();
+
+        // Wait for log handling threads to finish with timeout
+        let timeout = std::time::Duration::from_secs(5);
+        if let Some(handle) = self.stdout_handle.take() {
+            let _ = handle.join();
+        }
+        if let Some(handle) = self.stderr_handle.take() {
+            let _ = handle.join();
+        }
```
Comment (shekohex) [inline blueprint-test-utils/src/tangle/node.rs:384]: This says wait with timeout, but I do not see the timeout being used?

File: blueprint-test-utils/src/test_ext.rs [hunk 3]
```diff
@@ -171,13 +174,9 @@ pub async fn new_test_ext_blueprint_manager<
...
         None => {
-            let bytecode_hex = include_str!("../tnt-core/MasterBlueprintServiceManager.hex");
-            let mut raw_hex = bytecode_hex.replace("0x", "").replace("\n", "");
-            // fix odd length
-            if raw_hex.len() % 2 != 0 {
-                raw_hex = format!("0{}", raw_hex);
-            }
-            let bytecode = hex::decode(&raw_hex).expect("valid bytecode in hex format");
+            let bytecode = MASTER_BLUEPRINT_SERVICE_MANAGER.to_vec();
+            debug!("Using MBSM bytecode of length: {}", bytecode.len());
```
Comment (shekohex) [inline blueprint-test-utils/src/test_ext.rs:178]:
```suggestion
            trace!("Using MBSM bytecode of length: {}", bytecode.len());
```

File: blueprint-test-utils/src/test_ext.rs [hunk 4]
```diff
@@ -43,6 +44,7 @@ use gadget_sdk::subxt_core::config::Header;
 use gadget_sdk::utils::test_utils::get_client;
 use crate::tangle::node::SubstrateNode;
 use crate::tangle::transactions;
+use tnt_core_bytecode::bytecode::MASTER_BLUEPRINT_SERVICE_MANAGER;
```
Comment (shekohex) [inline blueprint-test-utils/src/test_ext.rs:47]: Where is this crate defined? forgot to push it?

**Review by shekohex — CHANGES_REQUESTED** [review]

**Review by shekohex — APPROVED** [review]

**Pull request merged**